### PR TITLE
LLVMContainerBase::cloneFiltered: fix dbg metadata

### PR DESCRIPTION
**Pull request by pfez**
Before this commit the cloneFiltered method was duplicating debug metadata on cloned functions, causing the Module not to verify after cloning.
The reason is a nasty combination of how llvm::cloneModule and llvm::Function::clearMetadata work internally:
- llvm::cloneModule already clones debug metadata, so the cloned functions have debug metadata
- llvm::Function::clearMetadata does not clear debug metadata
- so it's wrong to manually copy debug metadata from old functions to cloned functions

This commit fixes the problem by skipping the copy of the debug metadata alltoghether, since they're already in place.

### CODE REVIEW

**Comment by drblallo** [comment]
i think we may still need to manually copy debug metadata too if the destination function is a declaration. That too is kind of arbitrary and up to the particular version 
 of LLVM.

**Comment by pfez** [comment]
> i think we may still need to manually copy debug metadata too if the destination function is a declaration.

Why do you think so? AFAIK `cloneModule` already copies the debug metadata. Isn't this the case for the specific LLVM version we're using?

File: include/revng/Pipeline/LLVMContainer.h [hunk 1]
```diff
@@ -122,10 +122,18 @@ class LLVMContainerBase
       Other->clearMetadata();
       llvm::SmallVector<std::pair<unsigned, llvm::MDNode *>, 2> MDs;
       Function.getAllMetadata(MDs);
-      for (auto &MD : MDs)
-        Other->addMetadata(MD.first, *llvm::MapMetadata(MD.second, Map));
+      for (auto &[MDKindID, MetaData] : MDs) {
+        // CloneModule clone already clones debug metadata, and clearMetadata
+        // does not clear debug metadata, so we don't want to duplicate debug
+        // metadata because that causes the model to not verify.
```
Comment (aleclearmind) [inline include/revng/Pipeline/LLVMContainer.h:128]:
```suggestion
        // metadata because that causes the module to not verify.
```

**Comment by aleclearmind** [comment]
Please rebase, I get conflicts. We might already have a fix for this.

**Comment by pfez** [comment]
> Please rebase, I get conflicts. We might already have a fix for this.

Will do!

**Comment by aleclearmind** [comment]
Under which circumstance was this creating problems?

I'm fiddling with this piece of code.

**Comment by pfez** [comment]
> Under which circumstance was this creating problems?

I'm sorry but I don't remember exactly, it was already 2 months ago. I was debugging a pass downstream in `revng-c` and the module I was getting did not pass `verify` for this reason. So overall this may hamper debugging.

**Comment by aleclearmind** [comment]
I'll try this post LLVM 16 I guess.

**Comment by pfez** [comment]
Now that we've rebased on top of llvm-16 I'm retesting this. I'll mark this as draft until it's confirmed or it can be closed.

**Pull request closed without merging**